### ISSUE TITLE
Feature request: Check .env* files specially

### ISSUE THREAD

**Issue by jasonkarns** [koalaman/shellcheck#1275]
#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

(though searching existing issues was admittedly difficult because it's hard to match ".env" narrowly)

Suggestion: automatically disable [SC2034](https://github.com/koalaman/shellcheck/wiki/SC2034) ("foo appears unused. Verify it or export it.") for .env* files.

The .env file seems a common enough pattern, and there are loads of tools that use these files across multiple language communities. They are treated like shell files by most of these tools (supporting variable and command interpolation). However, they are rarely sourced and virtually never executed directly. Still, I find it valuable to run shellcheck on these files as a pseudo syntax highlighter and to catch variable/command interpolation issues, quoting problems, or bad variable assignments. Indeed, I believe most vim installations categorize .env files as SH filetypes by default, which means they are linted with shellcheck by default in most auto-linting configurations.

However, as these files are not truly shell files, they exist solely to set variables (but _not_ export them). It seems that shellcheck would do well to disable the unused/export variable check specifically for these files. (`.env*`) Adding the shell directive is easy enough. However, when developing across dozens of repos that each use this file, and knowing that each repo has probably half a dozen of these files (.env, .env.development, .env.production, .env.example, .env.local, etc), that is an awful lot of disable directives. It would be nice if that were automatic.

#### Here's a snippet or screenshot that shows the problem:

```sh
FOO=1
BAR=2
```

#### Here's what shellcheck currently says:

> foo appears unused. Verify it or export it.

#### Here's what I wanted or expected to see:

"no error"

relevant tooling: https://github.com/bkeepers/dotenv

**Comment by mightyiam** [comment]
Isn't this about more than specifically the `.env` pattern? It seems like shellcheck is not taking into consideration the feature of sourcing other scripts. The docs for this rule do recommend turning it off when the user knows that these variables are in use, but I'm not sure whether that falls under the stated category of "intentionally unused".

**Comment by jasonkarns** [comment]
@mightyiam I'm not sure I follow.

Are you suggesting that shellcheck doesn't account for variables used by sourced scripts?

If script `foo` sets variable `f` and sources script `bar` (which _uses_ `$f`), then you're correct:
`shellcheck foo` will report `f` as an unused variable.

However, if you pass the `-x` flag, then shellcheck will follow sourced files and will therefore find bar's use of `f` and not flag it as unused.

But really, that's besides the point for my scenario. I have .env* files, but if you use the `dotenv` gem for ruby or the corresponding npm package (or some other means of loading them), then the .env files will never be sourced directly by a shell script.

My purpose is to allow `shellcheck .env` to accurately report any errors _except_ `SC2034`, since that rule, nearly by definition and common use of .env files, will never be applicable.

**Comment by mightyiam** [comment]
`.env` files are a usage pattern. Seems out of the scope of a linter. Just use the disable comment in each of those `.env` files? That's probably what you're already doing.

**Comment by tonyseek** [comment]
The `.env` files looks like shell scripts but they are not shell scripts actually, if we accept the implementation of [forego](https://github.com/ddollar/forego).

```
$ cat .env
FOO=foo\ bar

$ cat env.sh
export FOO=foo\ bar
env | grep FOO

$ forego run env | grep FOO
FOO=foo\ bar

$ sh env.sh
FOO=foo bar
```

**Comment by cbows** [comment]
One solution would be to allow per file settings in `~/.shellcheckrc`. Instead of
```
$ cat ~/.shellcheckrc
disable=SC2034
```
we could have something like
```
$ cat ~/.shellcheckrc
.env* disable=SC2034
```
which disables `SC2034` only for files that match the pattern `.env*`.

AFAIK, this isn't possible currently though.
